### PR TITLE
[ticket/14134] Send warning notification PM in user's language.

### PR DESCRIPTION
**Pull request by lavigor**
PHPBB3-14134

### CODE REVIEW

**Comment by Nicofuma** [comment]
I wonder if this shouldn't use the notification system instead of the PM... anyway that wouldn't be for 3.1

File: phpBB/includes/mcp/mcp_warn.php [hunk 1]
```diff
@@ -530,12 +530,30 @@ function add_warning($user_row, $warning, $send_pm = true, $post_id = 0)
...
-		$user_row['user_lang'] = (file_exists($phpbb_root_path . 'language/' . $user_row['user_lang'] . "/mcp.$phpEx")) ? $user_row['user_lang'] : $config['default_lang'];
-		include($phpbb_root_path . 'language/' . basename($user_row['user_lang']) . "/mcp.$phpEx");
+		// Okay we need to get the reason from the warned users language
+		// Only load up the language pack if the language is different to the current one
+		if ($user_row['user_lang'] != $user->lang_name)
+		{
+			// Load up the language pack
+			$lang = array();
+
+			$user_row['user_lang'] = (file_exists($phpbb_root_path . 'language/' . $user_row['user_lang'] . "/mcp.$phpEx")) ? $user_row['user_lang'] : $config['default_lang'];
```
Comment (Nicofuma) [inline phpBB/includes/mcp/mcp_warn.php]: don't mix the notations please (avoid $var . 'text' . "$vartext$var" and choose one notation)

you can also move this check to the if.

File: phpBB/includes/mcp/mcp_warn.php [hunk 2]
```diff
@@ -530,12 +530,30 @@ function add_warning($user_row, $warning, $send_pm = true, $post_id = 0)
...
+			include($phpbb_root_path . 'language/' . basename($user_row['user_lang']) . "/mcp.$phpEx");
```
Comment (Nicofuma) [inline phpBB/includes/mcp/mcp_warn.php]: use exactly the same construction as in the check, use  `$user_row['user_lang']` in both cases

File: phpBB/includes/mcp/mcp_warn.php (hunk 1 above)
Comment (lavigor) [inline phpBB/includes/mcp/mcp_warn.php]: I haven't changed the previous code for this line. Do I need to make this adjustment?

File: phpBB/includes/mcp/mcp_warn.php (hunk 2 above)
Comment (lavigor) [inline phpBB/includes/mcp/mcp_warn.php]: > use  $user_row['user_lang']  in both cases

The same code is [here](https://github.com/phpbb/phpbb/blob/3.1.x/phpBB/includes/mcp/mcp_queue.php#L1253).
I treated it as an example.

File: phpBB/includes/mcp/mcp_warn.php [hunk 3]
```diff
@@ -530,12 +530,30 @@ function add_warning($user_row, $warning, $send_pm = true, $post_id = 0)
...
-		$user_row['user_lang'] = (file_exists($phpbb_root_path . 'language/' . $user_row['user_lang'] . "/mcp.$phpEx")) ? $user_row['user_lang'] : $config['default_lang'];
-		include($phpbb_root_path . 'language/' . basename($user_row['user_lang']) . "/mcp.$phpEx");
+		// Okay we need to get the reason from the warned users language
+		// Only load up the language pack if the language is different to the current one
+		if ($user_row['user_lang'] != $user->lang_name)
+		{
+			// Load up the language pack
+			$lang = array();
+
+			$user_row['user_lang'] = (file_exists($phpbb_root_path . 'language/' . basename($user_row['user_lang']) . "/mcp." . $phpEx)) ? $user_row['user_lang'] : $config['default_lang'];
```
Comment (Nicofuma) [inline phpBB/includes/mcp/mcp_warn.php]: I think you should move the file_exists check to the upper if

Comment (lavigor) [inline phpBB/includes/mcp/mcp_warn.php]: I thought about that, but the current approach is not the same.
No file in user's language => get the file in default language.
Default language may be different from the language of moderator who issued the warning.
That's why this line is not inside the if.

File: phpBB/includes/mcp/mcp_warn.php [hunk 4]
```diff
@@ -530,12 +530,30 @@ function add_warning($user_row, $warning, $send_pm = true, $post_id = 0)
 		include_once($phpbb_root_path . 'includes/functions_privmsgs.' . $phpEx);
 		include_once($phpbb_root_path . 'includes/message_parser.' . $phpEx);
 
-		$user_row['user_lang'] = (file_exists($phpbb_root_path . 'language/' . $user_row['user_lang'] . "/mcp.$phpEx")) ? $user_row['user_lang'] : $config['default_lang'];
-		include($phpbb_root_path . 'language/' . basename($user_row['user_lang']) . "/mcp.$phpEx");
+		// Okay we need to get the reason from the warned users language
+		// Only load up the language pack if the language is different to the current one
```
Comment (marc1706) [inline phpBB/includes/mcp/mcp_warn.php]: Please use a more objective comment. I'd recommend this:
```
Attempt to translate warning to language of user being warned if user's language differs from issuer's language
```

File: phpBB/includes/mcp/mcp_warn.php [hunk 5]
```diff
@@ -530,12 +530,30 @@ function add_warning($user_row, $warning, $send_pm = true, $post_id = 0)
...
+		if ($user_row['user_lang'] != $user->lang_name)
+		{
+			// Load up the language pack
```
Comment (marc1706) [inline phpBB/includes/mcp/mcp_warn.php]: Unneeded comment and actually wrong as you're only instantiating the array here.

File: phpBB/includes/mcp/mcp_warn.php [hunk 6]
```diff
@@ -530,12 +530,30 @@ function add_warning($user_row, $warning, $send_pm = true, $post_id = 0)
...
+			$lang = array();
+
+			$user_row['user_lang'] = (file_exists($phpbb_root_path . 'language/' . basename($user_row['user_lang']) . "/mcp." . $phpEx)) ? $user_row['user_lang'] : $config['default_lang'];
+			include($phpbb_root_path . 'language/' . basename($user_row['user_lang']) . "/mcp." . $phpEx);
+
+			$warn_pm_subject = $lang['WARNING_PM_SUBJECT'];
+			$warn_pm_body = sprintf($lang['WARNING_PM_BODY'], $warning);
+
+			unset($lang); // Free memory
```
Comment (marc1706) [inline phpBB/includes/mcp/mcp_warn.php]: Please also remove this comment. What this line of code does is very clear and there is no need in stating it with a comment.

**Comment by lavigor** [comment]
@marc1706 mentioned comments are not mine.
It's just a copypaste from [existing core code](https://github.com/phpbb/phpbb/blob/3.1.x/phpBB/includes/mcp/mcp_queue.php#L1265-L1279).

I guess comment adjustments require a separate issue and PR - `mcp_warn.php` is not the only place where fixes should be made.

**Comment by marc1706** [comment]
You're adding those comments here though. I don't care about changing old comments but I do care about adding new ones. Even if it's just a copy/paste from somewhere else, we shouldn't continue using improper comments just because we have used them before.

**Comment by lavigor** [comment]
@marc1706 Done.

**Comment by marc1706** [comment]
The "Okay ..." part is still in there

**Comment by Nicofuma** [comment]
@lavigor bump. Please remove the line starting with "Okay"

**Comment by lavigor** [comment]
@Nicofuma, done.

**Pull request merged**